### PR TITLE
Fabo/script runner

### PR DESCRIPTION
**Pull request by faboweb**
Server that runs scripts on http request with authentication

Missing: Switch in Polkadot Node Subscription to call remote instance

### CODE REVIEW

**Comment by faboweb** [comment]
🚀 Deployed on https://5ee5679f7e0d3b1b8d9f1a79--lunieio.netlify.app

**Comment by faboweb** [comment]
Test via:
curl --location --request POST 'https://scripts-staging.lunie.io/polkadotrewards' \
--header 'Authorization: ASK_FABO_FOR_CODE' \
--header 'Content-Type: application/json' \
--data-raw '{
    "era": 907
}'

File: scriptRunner/config.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+const secrets = require('../api/secrets')
+
+module.exports = {
+    authenticationToken: secrets.read('AUTHENTICATION_TOKEN') || process.env.AUTHENTICATION_TOKEN
```
Comment (Bitcoinera) [inline scriptRunner/config.js:4]: it would be better to call this SCRIPTRUNNER_AUTHENTICATION_TOKEN no? To be able to tell things apart

Comment (Bitcoinera) [reply scriptRunner/config.js:4]: ah, no, because it is another subrepo. Oh now I saw. Ignore my comment

File: scriptRunner/src/index.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        res
+            .status(403)
+            .send()
+        return
+    }
+    next()
+})
+
+
+app.post('/polkadotrewards', polkadotrewards)
```
Comment (Bitcoinera) [inline scriptRunner/src/index.js:30]: We could add an `app.get` here to query for the info..? Although the idea is to this from the API instance and querying against the Hasura DB right

**Comment by Bitcoinera** [comment]
Shouldn't this PR also include removing lines 154 - 180 in `polkadot-node-subscription.js`? I mean, we don't need to spawn `getOldPolkadotRewards` there anymore. We should instead query rewards from DB directly

File: scriptRunner/src/index.js (hunk 2 above)
Comment (faboweb) [reply scriptRunner/src/index.js:30]: what info would you like to get?

**Comment by faboweb** [comment]
Doesn't deploy to production yet

**Comment by Bitcoinera** [comment]
```
(node:5581) UnhandledPromiseRejectionWarning: FetchError: request to http://localhost:9000/polkadotrewards failed, reason: connect ECONNREFUSED 127.0.0.1:9000
    at ClientRequest.<anonymous> (/mnt/c/Users/usuario/lunie/lunie/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:189:13)
    at ClientRequest.EventEmitter.emit (domain.js:441:20)
    at Socket.socketErrorListener (_http_client.js:392:9)
    at Socket.emit (events.js:189:13)
    at Socket.EventEmitter.emit (domain.js:441:20)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
(node:5581) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
```

Probably a WSL problem?

File: api/lib/block-listeners/polkadot-node-subscription.js [hunk 3]
```diff
@@ -165,27 +165,24 @@ class PolkadotNodeSubscription {
...
+              config.scriptRunnerEndpoint
             )
-
-            if (code !== 0) {
-              Sentry.captureException(
-                new Error('getOldPolkadotRewardEras script failed')
-              )
-            }
-          })
+            fetch(`${config.scriptRunnerEndpoint}/polkadotrewards`, {
```
Comment (Bitcoinera) [inline api/lib/block-listeners/polkadot-node-subscription.js:173]:
```suggestion
            await fetch(`${config.scriptRunnerEndpoint}/polkadotrewards`, {
```

**Comment by Bitcoinera** [comment]
Correct me if I'm wrong, but we could actually delete te whole `updateRewards`, `eraRewardsExist` and `storeRewards` functions in `polkadot-node-subscription` (lines 227 - 264), right

**Comment by Bitcoinera** [comment]
Another observation: why I am not able to see any console logs from scriptRunner on the console? The ones from scheduleJob (the only ones)

image

File: api/lib/block-listeners/polkadot-node-subscription.js (hunk 3 above)
Comment (faboweb) [reply api/lib/block-listeners/polkadot-node-subscription.js:173]: why? we don't need to wait for this to finish

**Comment by faboweb** [comment]
> Another observation: why I am not able to see any console logs from scriptRunner on the console? The ones from scheduleJob (the only ones)

you got a 403 which means you don't have access. did you use the same authentication token in api and scriptRunner?

**Comment by Bitcoinera** [comment]
> you got a 403 which means you don't have access. did you use the same authentication token in api and scriptRunner?

Oh true, it is on the other computer

**Comment by Bitcoinera** [comment]
Getting undefineds

image

**Comment by Bitcoinera** [comment]
Then getting this other error after finished:

image

**Comment by faboweb** [comment]
Fixed era sending

**Comment by Bitcoinera** [comment]
And the other one, the unexpected token thing?

**Comment by Bitcoinera** [comment]
What about this one?

```
UnhandledPromiseRejectionWarning: FetchError: request to http://localhost:8080/v1/graphql failed, reason: connect ECONNREFUSED 127.0.0.1:8080
```

Seen in https://grafana.lunie.io/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bhost%3D%5C%22lunie-script-runner-staging%5C%22%7D%22%7D,%7B%22mode%22:%22Logs%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D

**Comment by Bitcoinera** [comment]
🤔 I don't see rewards from era 930 stored in staging-db

**Comment by faboweb** [comment]
in production-db there aren't either

**Comment by faboweb** [comment]
> What about this one?
> 
> ```
> UnhandledPromiseRejectionWarning: FetchError: request to http://localhost:8080/v1/graphql failed, reason: connect ECONNREFUSED 127.0.0.1:8080
> ```
> 
> Seen in https://grafana.lunie.io/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bhost%3D%5C%22lunie-script-runner-staging%5C%22%7D%22%7D,%7B%22mode%22:%22Logs%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D
> 
> Shouldn't it be `https://staging-db.lunie.io/v1/graphql` instead? Just asking

I fixed that one

**Pull request merged**